### PR TITLE
Hackathon: Post Reminders

### PR DESCRIPTION
**Pull request by agnivade**
This PR adds the post reminder backend work.

We add a new API endpoint via which a user can set a reminder for a post. An ephemeral message will be sent down the line to let the user know about the action. And then after the time is over, the system admin bot will send a DM message to the user about the reminder post.

```release-note
Added the server-side code for post reminders
```

### CODE REVIEW

File: api4/post_test.go [hunk 1]
```diff
@@ -3131,3 +3131,62 @@ func TestGetPostStripActionIntegrations(t *testing.T) {
...
+					require.NoError(t, err)
+
+					assert.Equal(t, model.PostTypeEphemeral, parsedPost.Type)
+					assert.Equal(t, th.BasicUser.Id, parsedPost.UserId)
+					assert.Equal(t, th.BasicPost.Id, parsedPost.RootId)
+
+					require.Equal(t, float64(targetTime), parsedPost.GetProp("target_time").(float64))
+					require.Equal(t, th.BasicPost.Id, parsedPost.GetProp("post_id").(string))
+					require.Equal(t, user.Username, parsedPost.GetProp("username").(string))
+					require.Equal(t, th.BasicTeam.Name, parsedPost.GetProp("team_name").(string))
```
Comment (isacikgoz) [inline api4/post_test.go:3183]: Shall we also return here?

File: app/post.go [hunk 2]
```diff
@@ -1750,6 +1750,121 @@ func (a *App) GetTopThreadsForUserSince(teamID, userID string, opts *model.Insig
...
 }
 
+func (a *App) SetPostReminder(postID, userID string, targetTime int64) *model.AppError {
+	// Store the reminder in the DB
+	reminder := &model.PostReminder{
+		PostId:     postID,
+		UserId:     userID,
+		TargetTime: targetTime,
+	}
+	err := a.Srv().Store.Post().SetPostReminder(reminder)
```
Comment (isacikgoz) [inline app/post.go:1760]: Would it be a good practice to check whether a post does exist or not before saving the reminder?

File: app/post.go [hunk 3]
```diff
@@ -1750,6 +1750,121 @@ func (a *App) GetTopThreadsForUserSince(teamID, userID string, opts *model.Insig
...
+					"Username": metadata.Username,
+				}),
+				Type:   model.PostTypeDefault,
+				UserId: systemBot.UserId,
+				Props: model.StringInterface{
+					"username": systemBot.Username,
+				},
+			}
+
+			if _, err := a.CreatePost(request.EmptyContext(), dm, ch, false, true); err != nil {
```
Comment (isacikgoz) [inline app/post.go:1860]: Wondering if we should delete the reminder from the db once we finished with it. Would it makes sense?

**Review by isacikgoz — COMMENTED** [review]
Looking forward to use this 🥇 did my first pass and have some questions.

File: app/post.go (hunk 2 above)
Comment (agnivade) [reply app/post.go:1760]: Yeah good point. We can use the `RowsAffected` field for this. I can make the changes.

File: app/post.go (hunk 3 above)
Comment (isacikgoz) [reply app/post.go:1860]: Ah, okay it escaped from me.

File: api4/post_test.go (hunk 1 above)
Comment (agnivade) [reply api4/post_test.go:3183]: Sure

**Review by isacikgoz — APPROVED** [review]
Noice 💯

**Comment by agnivade** [comment]
Oh I need to add the api reference as well.

File: i18n/en.json [hunk 4]
```diff
@@ -5903,6 +5903,14 @@
     "id": "app.post.update.app_error",
     "translation": "Unable to update the Post."
   },
+  {
+    "id": "app.post_reminder.app_error",
+    "translation": " "
+  },
+  {
+    "id": "app.post_reminder_dm",
+    "translation": "Hi there, you asked me to remind you about {{.SiteURL}}/{{.TeamName}}/pl/{{.PostId}}) by @{{.Username}}"
```
Comment (ashishbhate) [inline i18n/en.json:5912]: There's an extra parentheses after `{{.PostId}}`

```suggestion
    "translation": "Hi there, you asked me to remind you about {{.SiteURL}}/{{.TeamName}}/pl/{{.PostId}} by @{{.Username}}"
```

File: app/post.go [hunk 5]
```diff
@@ -1750,6 +1750,121 @@ func (a *App) GetTopThreadsForUserSince(teamID, userID string, opts *model.Insig
...
+}
+
+func (a *App) CheckPostReminders() {
+	systemBot, appErr := a.GetSystemBot()
+	if appErr != nil {
+		mlog.Error("Failed to get system bot", mlog.Err(appErr))
+		return
+	}
+
+	reminders, err := a.Srv().Store.Post().GetPostReminders(time.Now().UTC().Unix())
```
Comment (ashishbhate) [inline app/post.go:1813]: In `a.Srv().Store.Post().GetPostReminders(time.Now().UTC().Unix())` we delete reminders after getting them. 

If there there are any failures in this function after this point, then those reminders wont be sent and will be lost. 

Maybe we ought to consider spitting up the get and delete into separate operations? 

Non-blocking comment for this PR. We can handle this later.

**Review by ashishbhate — APPROVED** [review]
LGTM. 

Two small comments inline. One non-blocking discussion, other comment just highlights a small typo.

File: app/post.go (hunk 5 above)
Comment (agnivade) [reply app/post.go:1813]: Good point, I agree we can handle this later. I'll add a comment and create a ticket.

**Comment by ashishbhate** [comment]
The frontend component for this feature will take a lot more time to be ready. This PR is ready, and keeping it up to date is unnecessary work. 
Merging this PR. Any changes required by the frontend can come in follow-up PRs.

**Pull request merged**